### PR TITLE
handle S3 "KeyTooLongError" as PathNotFoundError in storage

### PR DESCRIPTION
**Pull request by syphar**
in [sentry we are seeing a new error from S3](https://rust-lang.sentry.io/issues/4864185388/events/57b8fec679f44b68b15feae0a6d5f766/?environment=production&project=5499376&query=is%3Aunresolved&referrer=next-event&statsPeriod=90d&stream_index=2): 

> ### Unhandled
> unhandled error (KeyTooLongError)
> ### ErrorMetadata
> Error { code: "KeyTooLongError", message: "Your key is too long", s3_extended_request_id: "8hzp1Kv7cGOsjW3EUaqBxaI8krQbNJYFX1UZdb5DzKsKpHG7B1K9rvRATlrT/MHw3kPtIbLIn9fEdGUrm85/Uw==", aws_request_id: "16PF5FB54MZZ081Q" }

it happens when the path in the requested URL is too long, and IMO should be seen as the same as a 404 from S3.

### CODE REVIEW

**Review by GuillaumeGomez — APPROVED** [review]

File: src/storage/s3.rs [hunk 1]
```diff
@@ -78,11 +91,11 @@ impl S3Backend {
         {
             Ok(_) => Ok(true),
             Err(SdkError::ServiceError(err))
-                if (matches!(err.err(), HeadObjectError::NotFound(_))
-                    || err.raw().status().as_u16() == http::StatusCode::NOT_FOUND.as_u16()) =>
+                if matches!(err.err(), HeadObjectError::NotFound(_)) =>
```
Comment (Nemo157) [inline src/storage/s3.rs:94]: Should this check be merged into `err_is_not_found` too?

File: src/storage/s3.rs [hunk 2]
```diff
@@ -21,6 +21,19 @@ use tracing::{error, warn};
...
 
+fn err_is_not_found<E>(err: &SdkError<E>) -> bool
+where
+    E: ProvideErrorMetadata,
+{
+    match err {
+        SdkError::ServiceError(err) => {
+            err.raw().status().as_u16() == http::StatusCode::NOT_FOUND.as_u16()
+        }
+        e if e.code() == Some("KeyTooLongError") => true,
```
Comment (Nemo157) [inline src/storage/s3.rs:32]: Does minio also return this error in some situations? Is it possible to write a testcase that triggers it?

Comment (syphar) [reply src/storage/s3.rs:32]: it seems to have it: 
https://github.com/minio/minio/blob/8cd967803cbccaac87bf71123ff86cff9281fc62/cmd/api-errors.go#L196 

let me check if I can trigger it

File: src/storage/s3.rs (hunk 1 above)
Comment (Nemo157) [reply src/storage/s3.rs:94]: Using just the `E: ProvideErrorMetadata`, I'd assume `e.code() == Some("NotFound")` would cover it (I tried checking what `code` would be returned for this case, but of course there's no easy way to find that in the aws-sdk source).

Comment (syphar) [reply src/storage/s3.rs:94]: Ah, now I know what you mean. Let me see

Comment (syphar) [reply src/storage/s3.rs]: @Nemo157 that was a nice idea, 

I adapted the logic, while finding the codes in the S3 SDK codebase & validating them with actual S3 requests.

**Review by Nemo157 — APPROVED** [review]

**Pull request merged**